packages/reactive_ble_mobile/ios/Classes/Prelude/CoreBluetooth+Extensions.swift

```diff
@@ -0,0 +1,19 @@
+import class CoreBluetooth.CBCharacteristic
+import class CoreBluetooth.CBService
+import class CoreBluetooth.CBPeripheral
+
+#if compiler(<5.5)
+extension CBCharacteristic {
+    @nonobjc
+    var service: CBService? {
```

Comment:
Both these properties are available since iOS 5.0. What you're trying to achieve with this?

Even if you have a reason to make this "shim", based on what (documentation) do you conclude that these objects actually have those named properties? This is to say it looks hacky / dirty.

Reply:
Hi! First of all, we need to figure out what exactly did Apple broke in Core Bluetooth APIs with Xcode 13. So they changed `- [CBCharacteristic service]` and `- [CBService peripheral]` semantic attribute from `unsafe` to `weak`. For Swift it requires now to add optional checks.
So what am I trying to achieve: 
- line 5: `compiler(<5.5)` means that included code should be compiled only in Xcode 13 and above (because only Xcode 13 do have Swift compiler version 5.5). Swift compiler version depends on Xcode version even so the project is compiled with Swift 4.2;
- line 9: #function == "service", I don't mind changing it to string literal;
- line 16: #function == "peripheral", I don't mind changing it to string literal;
- headers for `CBCharacteristic` class [here](https://developer.limneos.net/?ios=11.0&framework=CoreBluetooth.framework&header=CBCharacteristic.h) and `CBService` class [here](https://developer.limneos.net/?ios=11.0&framework=CoreBluetooth.framework&header=CBService.h) guarantee that above-mentioned classes will have required named properties in runtime;
- `@nonobjc` and internal declaration guarantee that we will not affect API behaviour anywhere outside our framework;
- converting properties from `unsafe` to `weak` is _okay_ since Apple intended them to be unsafe before Xcode 13;
- this fix doesn't required any changes in existing codebase.
Thus with this "shim" we have APIs which works identically both in Xcode 12 and Xcode 13, which I assume is pretty sweet.

Reply:
@ky1vstar That's a pretty good explanation, I appreciate your effort. Some feedback:

> line 9: #function == "service", I don't mind changing it to string literal;
> line 16: #function == "peripheral", I don't mind changing it to string literal;

I actually like the current approach. Let's keep that as is.

> headers for `CBCharacteristic` class [here](https://developer.limneos.net/?ios=11.0&framework=CoreBluetooth.framework&header=CBCharacteristic.h) and `CBService` class [here](https://developer.limneos.net/?ios=11.0&framework=CoreBluetooth.framework&header=CBService.h) guarantee that above-mentioned classes will have required named properties in runtime;

You need to refer to the official documentation to derive any guarantees. That dump just proves certain properties of the _implementation_ in a particular OS build. Only the documented _public interface_ is reliable, not undocumented implementation details.

In this case, though, the subject properties _are_ a part of the public interface, so the official documentation says something very similar:

- [CBCharacteristic.service](https://developer.apple.com/documentation/corebluetooth/cbcharacteristic/1518728-service)
- [CBService.peripheral](https://developer.apple.com/documentation/corebluetooth/cbservice/1434334-peripheral)

> converting properties from `unsafe` to `weak` is okay since Apple intended them to be unsafe before Xcode 13;

You're not converting them to `weak`, though. The "shim" properties are now declared with a strong reference type. I'd rather prefer them to mimic the newer interface and use `weak` qualifier.

---

You see, the `unsafe` vs. `weak` is a subtle detail. This is worth putting in a comment or a doc-comment around the shim properties, otherwise it will be nigh impossible to figure out why that is needed some time later.

Please, reflect in the comments or doc-comments that:

- this code "backports" the API changes made in Xcode 13 so that they are available in Xcode 12
- namely, this code changes the type of the properties from unsafe non-optional to weak (optional)

---

So,

- [ ] with the addition of some comments or doc-comments
- [ ] and using `weak` qualifier in the shim properties declaration

this patch seems acceptable to me after your explanation.

Reply:
@werediver I'll provide comment a little bit later, but I do already user `weak` (optional) qualifier, don't I?

Reply:
@werediver thank you, I'll make appropriate changes.

Reply:
@werediver done